airflow/datasets/manager.py

```diff
@@ -58,12 +58,55 @@ class DatasetManager(LoggingMixin):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
-    def create_datasets(self, dataset_models: list[DatasetModel], session: Session) -> None:
+    def create_datasets(self, datasets: list[Dataset], *, session: Session) -> list[DatasetModel]:
         """Create new datasets."""
-        for dataset_model in dataset_models:
-            session.add(dataset_model)
-        for dataset_model in dataset_models:
-            self.notify_dataset_created(dataset=Dataset(uri=dataset_model.uri, extra=dataset_model.extra))
+
+        def _add_one(dataset: Dataset) -> DatasetModel:
+            model = DatasetModel.from_public(dataset)
+            session.add(model)
+            return model
+
+        models = [_add_one(d) for d in datasets]
+        for dataset in datasets:
+            self.notify_dataset_created(dataset=dataset)
```

Comment:
Why not doing both operations in one for loop? We loop through the datasets twice in a row

Reply:
I don’t know TBH, this was how it was done before.

#34418 maybe @yermalov-here has some reasons?

Reply:
One possibility is maybe we want to ensure all datasets are in the session before we call the hooks?

Reply:
We had two choices when we did this (to flush or not to flush before calling the hook). Was there a particular reason one was chosen instead of the other?

Reply:
I am not aware of the exact implications as my knowledge in this area is limited. I think the main reason was consistency: all the dataset hooks are called after flushing.

Reply:
I let you decide whether you want to keep it in 2 different loops and to merge them into one :) My point was mostly related to coding style but if there is a valid reason why we need 2 loops, I am fine

Reply:
I think I’ll add back the flush instead. Flushing more often shouldn’t affect performance much (we need to write to the database _eventually_, it’s just the question when). But I’ll also mention this in the docstring so we have the liberty to remove the flush if needed in the future.

Reply:
On second though, maybe not…? Flushing does not actually make the objects visible (because transaction) so the listener hook still cannot “see” the object in the database (because it does not have access to the session used to insert the object). So the additional flush in the middle of the function does not actually affect anything in the listener. I’ll combine the loops. In a follow-up PR I’ll maybe also refactor `register_dataset_change`